src/sentry/sentry_metrics/last_seen_updater.py

```diff
@@ -1,26 +1,78 @@
-from typing import Mapping, Optional, Union
+import functools
+import random
+from datetime import timedelta
+from typing import Any, Mapping, Optional, Set, Union
 
-from arroyo import Topic
+from arroyo import Message, Topic
 from arroyo.backends.kafka import KafkaConsumer, KafkaPayload
 from arroyo.processing import StreamProcessor
 from arroyo.processing.strategies import ProcessingStrategy
 from arroyo.processing.strategies.streaming import KafkaConsumerStrategyFactory
 from arroyo.processing.strategies.streaming.factory import StreamMessageFilter
+from django.utils import timezone
 
+from sentry import options
+from sentry.sentry_metrics.indexer.base import FetchType
+from sentry.sentry_metrics.indexer.models import StringIndexer
 from sentry.sentry_metrics.multiprocess import get_config, logger
+from sentry.utils import json
 
+MAPPING_META = "mapping_meta"
 
-class LastSeenUpdaterMessageFilter(StreamMessageFilter[KafkaPayload]):  # type: ignore
-    def should_drop(self, message: KafkaPayload) -> bool:
-        return False
 
+@functools.lru_cache(maxsize=10)
+def get_metrics():  # type: ignore
+    from sentry.utils import metrics
 
-class LastSeenUpdaterCollector(ProcessingStrategy[int]):  # type: ignore
-    def __init__(self) -> None:
-        self.__counter = 0
+    return metrics
 
-    def submit(self, message: int) -> None:
-        self.__counter += message
+
+class LastSeenUpdaterMessageFilter(StreamMessageFilter[Message[KafkaPayload]]):  # type: ignore
+    def __init__(self, metrics: Any) -> None:
+        self.__metrics = metrics
+
+    # We want to ignore messages where the mapping_sources header is present
+    # and does not contain the DB_READ ('d') character (this should be the vast
+    # majority of messages).
+    def should_drop(self, message: Message[KafkaPayload]) -> bool:
+        feature_enabled: float = options.get("sentry-metrics.last-seen-updater.accept-rate")
+        if random.random() > feature_enabled:
+            self.__metrics.incr("last_seen_updater.accept_rate_bypass")
+            return True
+
+        header_value: Optional[str] = next(
+            (
+                str(header[1])
+                for header in message.payload.headers
+                if header[0] == "mapping_sources"
+            ),
+            None,
+        )
+
+        if not header_value:
+            self.__metrics.incr("last_seen_updater.header_not_present")
```

Comment:
The header isn't added until this PR is fully deployed (see the change to multiprocess.py).

In general I think we want to know if somebody unknowingly drops that because it will cause performance to degrade. Alternately it could be an error case but I wasn't sure if we wanted to do that since we can still process the message, it just requires more CPU than if we could determine it was skippable without parsing JSON